docs/proposals/remote-artifacts.md

```diff
@@ -0,0 +1,267 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: remote-artifacts
+authors:
+  - "@otaviof"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-10-02
+last-updated:  2020-10-02
+status: provisional
+---
+
+Shipwright Remote Artifacts
+---------------------------
+
+# Summary
+
+Remote artifacts are often dependencies of the software building process, and hence are also
+required when dealing with container images. This enhancement proposal focuses on adding a build
+artifacts abstraction to Shipwright's Operator.
+
+# Motivation
+
+Give Shipwright's operator broader build use-case support by enhancing its capabilities to include
+the concept of Artifacts. In other words, remote entities that will be available for the image
+build process.
+
+End users will be allowed to declare artifacts that represent remote dependencies alongside
+Builds, and easily link them together. For example, a Java application which downloads certain
+jars into the classpath, a Node.js application adding external images, and so forth.
+
+## Goals
+
+*   Provide means to declare remote artifacts, dependencies that can be employed on builds;
+*   Create the mechanism to download and prepare remote artifacts for builds;
+
+## Non-Goals
+
+*   Automate the upload of local artifacts into the cluster;
+*   Manage remote artifacts;
+*   Amend container images created by the given `BuildStrategy`;
+*   Walking a remote directory tree, similarly to a `git checkout`;
+
+# Proposal
+
+The enhancement proposal is centered around the idea of declaring external dependencies, here
+called "artifacts", and being able to use them during the container image building process.
+
+## Build Artifacts
+
+The remote artifacts will be directly expressed in the body of a `Build` resource, as the
+following example:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: nexus-jar
+spec:
+  artifacts:
+    - name: main-jar
+      url: https://nexus.company.com/app/main.jar
+      path: $(workspace)/classpath/main.jar
+```
+
+A new attribute, .spec.artifacts will be added, containing a slice of types with:
+
+*   **Name**: The actual name of the artifact, optional;
```

Comment:
I would make `Name` required - if artifacts and/or sources are going to be a list, we need a means of doing strategic merge patching [1]. Per discussions I've had with sig-apimachinery members, having the merge patch key be required is strongly recommend.

https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch

Reply:
Indeed, that's a good addition.